saleor/warehouse/reservations.py

```diff
@@ -0,0 +1,160 @@
+from collections import defaultdict, namedtuple
+from typing import TYPE_CHECKING, Dict, Iterable, List, cast
+
+from django.db import transaction
+from django.db.models import Sum
+
+from ..core.exceptions import InsufficientStock, InsufficientStockData
+from ..product.models import ProductVariant
+from .models import Allocation, Reservation, Stock
+
+if TYPE_CHECKING:
+    from ..checkout.models import CheckoutLine
+
+StockData = namedtuple("StockData", ["pk", "quantity"])
+
+
+@transaction.atomic
+def reserve_stocks(
+    checkout_lines: Iterable["CheckoutLine"],
+    country_code: str,
+    channel_slug: str,
+    *,
+    replace=True,
+):
+    """Reserve stocks for given `checkout_lines` in given country."""
+    # Reservation is only applied to checkout lines with variants with track inventory
+    # set to True
+    checkout_lines = get_checkout_lines_to_reserve(checkout_lines)
+    if not checkout_lines:
+        return
+
+    variants = [line.variant for line in checkout_lines]
+
+    stocks = list(
+        Stock.objects.select_for_update(of=("self",))
+        .get_variants_stocks_for_country(country_code, channel_slug, variants)
+        .order_by("pk")
+        .values("id", "product_variant", "pk", "quantity")
+    )
+    stocks_id = (stock.pop("id") for stock in stocks)
+
+    quantity_allocation_list = list(
+        Allocation.objects.filter(
+            stock_id__in=stocks_id,
+            quantity_allocated__gt=0,
+        )
+        .values("stock")
+        .annotate(quantity_allocated_sum=Sum("quantity_allocated"))
+    )
+    quantity_allocation_for_stocks: Dict = defaultdict(int)
+    for allocation in quantity_allocation_list:
+        quantity_allocation_for_stocks[allocation["stock"]] += allocation[
+            "quantity_allocated_sum"
+        ]
+
+    quantity_reservation_list = list(
+        Reservation.objects.filter(
+            stock_id__in=stocks_id,
+            quantity_reserved__gt=0,
+        )
+        .not_expired()
+        .exclude_checkout_lines(checkout_lines)
+        .values("stock")
+        .annotate(quantity_reserved_sum=Sum("quantity_reserved"))
+    )  # type: ignore
+    quantity_reservation_for_stocks: Dict = defaultdict(int)
+    for reservation in quantity_reservation_list:
+        quantity_reservation_for_stocks[allocation["stock"]] += reservation[
```

Comment:
We should cover this with tests. I think it's enough to add a test for creating reservations for lines that contain variants with non 0 allocations and non 0 reservations.